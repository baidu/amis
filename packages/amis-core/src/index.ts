export * from './abc';

export function hello() {
  return 'world';
}
