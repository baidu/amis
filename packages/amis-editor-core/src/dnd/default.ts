/**
 * 默认的拖拽模式实现。
 */
import {animation} from 'amis';
import findIndex from 'lodash/findIndex';
import {EditorDNDManager} from '.';
import {renderThumbToGhost} from '../component/factory';
import {EditorNodeType} from '../store/node';
import {translateSchema} from '../util';
import {DNDModeInterface} from './interface';

export class DefaultDNDMode implements DNDModeInterface {
  readonly dndContainer: HTMLElement; // 记录当前拖拽区域
  dropBeforeId?: string;
  constructor(readonly dnd: EditorDNDManager, readonly region: EditorNodeType) {
    // 初始化时，默认将元素所在区域设置为当前拖拽区域
    this.dndContainer = this.dnd.store
      .getDoc()
      .querySelector(
        `[data-region="${region.region}"][data-region-host="${region.id}"]`
      ) as HTMLElement;
  }

  /**
   * 记录上次交换时的鼠标位置。
   */
  exchangeX: number = 0;
  exchangeY: number = 0;

  /**
   * 首次拖入，把 ghost 插入进来。让用户有个直观感受。
   * @param e
   * @param ghost
   */
  enter(e: DragEvent, ghost: HTMLElement) {
    const dragEl = this.dnd.dragElement;
    // 如果就在移动的元素所在的区域，那么把 ghost 放到原来的位置。
    if (dragEl && dragEl.closest('[data-region]') === this.dndContainer) {
      const list: Array<any> = Array.isArray(this.region.schema)
        ? this.region.schema
        : [];
      const child = this.getChild(this.dndContainer, dragEl);
      const id = dragEl.getAttribute('data-editor-id')!;
      const idx = findIndex(list, (item: any) => item.$$id === id);
      if (~idx && list[idx + 1]) {
        this.dropBeforeId = list[idx + 1].$$id;
      }
      this.dndContainer.insertBefore(ghost, child);

      let innerHTML = dragEl.outerHTML
        .replace('ae-is-draging', '')
        // .replace(/\bdata\-editor\-id=('|").+?\1/g, '');
        // 上面那个会让 fis 编译的时候把后续的 require 语句都给忽略掉了。
        .replace(/\bdata\-editor\-id=(?:'.+?'|".+?")/g, '');
      // bca-disable-next-line
      ghost.innerHTML = innerHTML;
    } else {
      const manager = this.dnd.manager;
      const store = manager.store;
      renderThumbToGhost(
        ghost,
        this.region,
        translateSchema(store.dragSchema),
        manager
      );
      this.dndContainer.appendChild(ghost);
    }
  }

  /**
   * 拖出去了，就移除 ghost
   * @param e
   * @param ghost
   */
  leave(e: DragEvent, ghost: HTMLElement) {
    this.dndContainer.removeChild(ghost);
  }

  over(e: DragEvent, ghost: HTMLElement) {
    const target = this.getTarget(e);
    const wrapper = this.dndContainer;
    const elemSchema = this.region.schema;
    const list: Array<any> = Array.isArray(elemSchema) ? elemSchema : [];
    const dx = e.clientX - this.exchangeX;
    const dy = e.clientY - this.exchangeY;
    const vertical = Math.abs(dy) > Math.abs(dx);

    if (target && !animation.animating) {
      const targetId = target.getAttribute('data-editor-id')!;
      const targetChild = this.getChild(wrapper, target!);
      const idx = findIndex(list, (item: any) => item.$$id === targetId);

      const originIdx = Array.prototype.indexOf.call(wrapper.children, ghost);
      const targetIdx = Array.prototype.indexOf.call(
        wrapper.children,
        targetChild
      );

      if (
        ~originIdx &&
        originIdx > targetIdx &&
        (!this.exchangeY || dy < 0 || dx < 0)
      ) {
        // 原来在后面，移动到前面

        this.exchangeX = e.clientX;
        this.exchangeY = e.clientY;
        this.dropBeforeId = list[idx]?.$$id;

        if (originIdx !== targetIdx - 1) {
          animation.capture(wrapper);
          wrapper.insertBefore(ghost, targetChild);
          animation.animateAll();
        }
      } else if (
        ~originIdx &&
        originIdx < targetIdx &&
        (!this.exchangeY || dy > 0 || dx > 0)
      ) {
        // 原来在前面，移动到后面

        this.exchangeX = e.clientX;
        this.exchangeY = e.clientY;
        if (list[idx + 1]) {
          this.dropBeforeId = list[idx + 1]?.$$id;
        } else {
          delete this.dropBeforeId;
        }

        if (originIdx !== targetIdx + 1) {
          animation.capture(wrapper);
          wrapper.insertBefore(ghost, targetChild.nextSibling);
          animation.animateAll();
        }
      }
    }

    if (ghost.parentNode !== wrapper) {
      delete this.dropBeforeId;
      animation.capture(wrapper);
      wrapper.appendChild(ghost);
      animation.animateAll();
    }
  }

  /**
   * 返回个相对位置，如果没有数据会插入到结尾。
   */
  getDropBeforeId() {
    return this.dropBeforeId;
  }

  /**
   * 获取当时拖动到了哪个节点上面。
   */
  getTarget(e: DragEvent) {
    let target = (e.target as HTMLElement).closest(
      '[data-editor-id]'
    ) as HTMLElement;

    while (target) {
      const region = target.parentElement?.closest('[data-region]');

      if (region === this.dndContainer) {
        const renderer = target.getAttribute('data-renderer');
        if (renderer === 'grid') {
          // grid 组件中的分栏的子栏也可以拖入 选中分栏组件同级组件拖动时有问题 兼容一下
          return target.parentElement;
        } else {
          return target;
        }
      }

      target =
        (target.parentElement?.closest('[data-editor-id]') as HTMLElement) ||
        null;
    }

    return null;
  }

  /**
   * 获取区域的直接孩子，因为有时候会在孩子的孩子里面。
   * 但是插入 ghost 的相对位置，insertBefore 只能是当前孩子。
   * @param dom
   * @param descend
   */
  getChild(dom: HTMLElement, descend: HTMLElement) {
    let child = descend;

    while (child) {
      if (child.parentElement === dom) {
        break;
      }

      child = child.parentElement!;
    }

    return child;
  }

  /**
   * 销毁
   */
  dispose() {
    delete this.dropBeforeId;
  }
}
