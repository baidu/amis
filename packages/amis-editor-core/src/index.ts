/**
 * @file 总入口
 */
import 'amis';
import './locale/index';
import Editor from './component/Editor';
import './component/ClassNameControl';

import './plugin/AvailableRenderers';
import './plugin/BasicToolbar';
import './plugin/Code';
import './plugin/DataDebug';
import './plugin/ErrorRenderer';
// import './plugin/Name';
import './plugin/Outline';
import './plugin/Unknown';

import * as utils from './util';
export * from './util';
export * from './tpl';
export * from './manager';
export * from './plugin';
export * from './icons/index';
export * from './mocker';
import {BasicEditor, RendererEditor} from './compat';
import MiniEditor from './component/MiniEditor';
import CodeEditor from './component/Panel/AMisCodeEditor';
import IFramePreview from './component/IFramePreview';
import SearchPanel from './component/base/SearchPanel';
import {VRenderer} from './component/VRenderer';
import {RegionWrapper} from './component/RegionWrapper';
import {mapReactElement} from './component/factory';
import type {EditorNodeType, EditorNodeSnapshot} from './store/node';
import {ContainerWrapper} from './component/ContainerWrapper';
import type {EditorStoreType} from './store/editor';
import {AvailableRenderersPlugin} from './plugin/AvailableRenderers';
import ShortcutKey from './component/base/ShortcutKey';
import WidthDraggableContainer from './component/base/WidthDraggableContainer';

export const version = '__buildVersion';

export default Editor;

export {
  Editor,
  MiniEditor,
  utils,
  mapReactElement,
  RendererEditor,
  BasicEditor,
  CodeEditor,
  VRenderer,
  RegionWrapper,
  IFramePreview as IFrameEditor,
  SearchPanel,
  EditorNodeType,
  EditorNodeSnapshot,
  EditorStoreType,
  ContainerWrapper,
  AvailableRenderersPlugin,
  ShortcutKey,
  WidthDraggableContainer
};
