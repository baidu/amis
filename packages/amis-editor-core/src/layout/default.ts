import {LayoutInterface} from './interface';

export default class DefaultLayout implements LayoutInterface {}
