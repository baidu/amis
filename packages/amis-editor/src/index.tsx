import 'amis';
import './locale/index';
export * from 'amis-editor-core';
export * from './builder';
import './tpl/index';
export * from './plugin';

import './renderer/OptionControl';
import './renderer/NavSourceControl';
import './renderer/NavBadgeControl';
import './renderer/NavDefaultActive';
import './renderer/MapSourceControl';
import './renderer/TimelineItemControl';
import './renderer/APIControl';
import './renderer/APIAdaptorControl';
import './renderer/ValidationControl';
import './renderer/ValidationItem';
import './renderer/SwitchMoreControl';
import './renderer/StatusControl';
import './renderer/FormulaControl';
import './renderer/ExpressionFormulaControl';
import './renderer/textarea-formula/TextareaFormulaControl';
import './renderer/TplFormulaControl';
import './renderer/DateShortCutControl';
import './renderer/BadgeControl';
import './renderer/style-control/BoxModel';
import './renderer/style-control/Font';
import './renderer/style-control/Border';
import './renderer/style-control/BoxShadow';
import './renderer/style-control/Background';
import './renderer/style-control/Display';
import './renderer/style-control/InsetBoxModel';
import './renderer/RangePartsControl';
import './renderer/DataBindingControl';
import './renderer/DataMappingControl';
import './renderer/DataPickerControl';
import './renderer/FeatureControl';
import './renderer/event-control/index';
import './renderer/TreeOptionControl';
import './renderer/TransferTableControl';
import './renderer/style-control/ThemeCssCode';
import './renderer/ButtonGroupControl';
import './renderer/FlexSettingControl';
import './renderer/FieldSetting';
import './renderer/TableColumnWidthControl';
import './renderer/crud2-control/CRUDColumnControl';
import './renderer/crud2-control/CRUDToolbarControl';
import './renderer/crud2-control/CRUDFiltersControl';
import './renderer/InputRangeValueControl';
import './renderer/FunctionEditorControl';
import './renderer/ListItemControl';

import 'amis-theme-editor/lib/locale/zh-CN';
import 'amis-theme-editor/lib/locale/en-US';
import 'amis-theme-editor/lib/renderers/Border';
import 'amis-theme-editor/lib/renderers/ColorPicker';
import 'amis-theme-editor/lib/renderers/Font';
import 'amis-theme-editor/lib/renderers/PaddingAndMargin';
import 'amis-theme-editor/lib/renderers/Radius';
import 'amis-theme-editor/lib/renderers/Shadow';
import 'amis-theme-editor/lib/renderers/Size';
import 'amis-theme-editor/lib/renderers.css';

export * from './component/BaseControl';
export * from './icons/index';
