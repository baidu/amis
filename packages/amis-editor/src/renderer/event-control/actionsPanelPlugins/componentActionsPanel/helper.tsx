import {JSONGetById, EditorManager} from 'amis-editor-core';
import {DataSchema} from 'amis-core';
import CmptActionSelect from '../../comp-action-select';

// 下拉展示可赋值属性范围
export const SELECT_PROPS_CONTAINER = ['form'];

export const renderCmptActionSelect = (
  componentLabel: string,
  required: boolean,
  onChange?: (value: string, oldVal: any, data: any, form: any) => void,
  hideAutoFill?: boolean,
  manager?: EditorManager
) => {
  return [
    ...renderCmptSelect(
      componentLabel || '选择组件',
      true,
      async (value: string, oldVal: any, data: any, form: any) => {
        // 获取组件上下文
        if (form.data.__nodeId) {
          if (form.data.actionType === 'setValue') {
            // todo:这里会闪一下，需要从amis查下问题
            form.setValueByName('args.value', []);
            form.setValueByName('args.__comboType', undefined);
            form.setValueByName('args.__valueInput', undefined);
            form.setValueByName('args.__containerType', undefined);

            if (SELECT_PROPS_CONTAINER.includes(form.data.__rendererName)) {
              const contextSchema: any = await form.data.getContextSchemas?.(
                form.data.__nodeId,
                true
              );

              const dataSchema = new DataSchema(contextSchema || []);
              const variables = dataSchema?.getDataPropsAsOptions() || [];
              form.setValueByName(
                '__setValueDs',
                variables.filter(item => item.value !== '$$id')
              );
            } else {
              form.setValueByName('__setValueDs', []);
            }
          }
        }
        form.setValueByName('groupType', '');
        onChange?.(value, oldVal, data, form);
      },
      hideAutoFill
    ),
    {
      type: 'input-text',
      name: '__cmptId',
      mode: 'horizontal',
      size: 'lg',
      required: true,
      label: '组件id',
      visibleOn:
        'this.componentId === "customCmptId" && this.actionType === "component"',
      onChange: async (value: string, oldVal: any, data: any, form: any) => {
        let schema = JSONGetById(manager!.store.schema, value, 'id');
        if (schema) {
          form.setValues({
            __rendererName: schema.type
          });
        } else {
          form.setValues({
            __rendererName: ''
          });
        }
      }
    },
    {
      asFormItem: true,
      label: '组件动作',
      name: 'groupType',
      mode: 'horizontal',
      required: true,
      visibleOn: 'this.actionType === "component"',
      component: CmptActionSelect,
      description: '${__cmptActionDesc}'
    }
  ];
};

export const renderCmptSelect = (
  componentLabel: string,
  required: boolean,
  onChange?: (value: string, oldVal: any, data: any, form: any) => void,
  hideAutoFill?: boolean
) => [
  {
    type: 'tree-select',
    name: 'componentId',
    label: componentLabel || '选择组件',
    showIcon: false,
    searchable: true,
    required,
    selfDisabledAffectChildren: false,
    size: 'lg',
    source: '${__cmptTreeSource}',
    mode: 'horizontal',
    autoFill: {
      __isScopeContainer: '${isScopeContainer}',
      ...(hideAutoFill
        ? {}
        : {
            __rendererLabel: '${label}',
            __rendererName: '${type}',
            __nodeId: '${id}',
            __nodeSchema: '${schema}'
          })
    },
    onChange: async (value: string, oldVal: any, data: any, form: any) => {
      onChange?.(value, oldVal, data, form);
    }
  }
];

export const renderCmptIdInput = (
  onChange?: (value: string, oldVal: any, data: any, form: any) => void
) => {
  return {
    type: 'input-text',
    name: '__cmptId',
    mode: 'horizontal',
    size: 'lg',
    required: true,
    label: '组件id',
    visibleOn: 'this.componentId === "customCmptId"',
    onChange: async (value: string, oldVal: any, data: any, form: any) => {
      onChange?.(value, oldVal, data, form);
    }
  };
};
