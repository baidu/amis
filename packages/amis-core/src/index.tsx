/** @license amis v@version
 *
 * Copyright Baidu
 *
 * This source code is licensed under the Apache license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  Renderer,
  getRendererByName,
  getRenderers,
  registerRenderer,
  unRegisterRenderer,
  resolveRenderer,
  filterSchema,
  clearStoresCache,
  updateEnv,
  stores,
  defaultOptions,
  addSchemaFilter,
  extendDefaultEnv
} from './factory';
import type {RenderOptions, RendererConfig, RendererProps} from './factory';
import './polyfills';
import './renderers/builtin';
import './renderers/register';
export * from './utils/index';
export * from './types';
export * from './store';
import * as utils from './utils/helper';
import {getEnv} from 'mobx-state-tree';

import {RegisterStore, RendererStore} from './store';
import type {IColumn, IColumn2, IRow, IRow2} from './store';
import {
  setDefaultLocale,
  getDefaultLocale,
  makeTranslator,
  register as registerLocale,
  extendLocale,
  removeLocaleData,
  localeable
} from './locale';
import type {LocaleProps, TranslateFn} from './locale';

import Scoped, {ScopedContext, filterTarget} from './Scoped';
import type {ScopedComponentType, IScopedContext} from './Scoped';

import {
  classnames,
  getClassPrefix,
  setDefaultTheme,
  theme,
  getTheme,
  themeable,
  makeClassnames
} from './theme';
import type {ClassNamesFn, ThemeProps} from './theme';
const classPrefix = getClassPrefix();

export * from './actions';
import FormItem, {
  FormItemWrap,
  registerFormItem,
  getFormItemByName
} from './renderers/Item';
import type {
  FormBaseControl,
  FormControlProps,
  FormItemProps
} from './renderers/Item';
import {OptionsControl, registerOptionsControl} from './renderers/Options';
import type {OptionsControlProps} from './renderers/Options';
import type {FormOptionsControl} from './renderers/Options';
import {Schema} from './types';
import ScopedRootRenderer, {addRootWrapper, RootRenderProps} from './Root';
import {envOverwrite} from './envOverwrite';
import {EnvContext} from './env';
import type {RendererEnv} from './env';
import React from 'react';
import {
  evaluate,
  evaluateForAsync,
  Evaluator,
  AsyncEvaluator,
  extendsFilters,
  filters,
  getFilters,
  lexer,
  parse,
  registerFilter,
  registerFunction
} from 'amis-formula';
import type {FilterContext} from 'amis-formula';
import LazyComponent from './components/LazyComponent';
import Overlay from './components/Overlay';
import PopOver from './components/PopOver';
import {FormRenderer} from './renderers/Form';
import type {FormHorizontal, FormSchemaBase} from './renderers/Form';
import {
  enableDebug,
  disableDebug,
  promisify,
  replaceText,
  wrapFetcher
} from './utils/index';
import type {OnEventProps} from './utils/index';
import {valueMap as styleMap} from './utils/style-helper';
import {RENDERER_TRANSMISSION_OMIT_PROPS} from './SchemaRenderer';
import type {IItem} from './store/list';
import CustomStyle from './components/CustomStyle';
import {StatusScoped} from './StatusScoped';

// @ts-ignore
export const version = '__buildVersion';

export {
  clearStoresCache,
  updateEnv,
  Renderer,
  RendererProps,
  RenderOptions,
  RendererEnv,
  EnvContext,
  RegisterStore,
  FormItem,
  FormItemWrap,
  FormItemProps,
  OptionsControl,
  FormRenderer,
  FormHorizontal,
  // 其他功能类方法
  utils,
  getRendererByName,
  registerRenderer,
  unRegisterRenderer,
  getRenderers,
  registerFormItem,
  getFormItemByName,
  registerOptionsControl,
  resolveRenderer,
  filterSchema,
  Scoped,
  ScopedContext,
  IScopedContext,
  StatusScoped,
  setDefaultTheme,
  theme,
  themeable,
  ThemeProps,
  getTheme,
  classPrefix,
  getClassPrefix,
  classnames,
  makeClassnames,
  // 多语言相关
  getDefaultLocale,
  setDefaultLocale,
  registerLocale,
  makeTranslator,
  extendLocale,
  removeLocaleData,
  localeable,
  LocaleProps,
  TranslateFn,
  ClassNamesFn,
  // amis-formula 相关
  parse,
  lexer,
  Evaluator,
  AsyncEvaluator,
  FilterContext,
  filters,
  getFilters,
  registerFilter,
  extendsFilters,
  registerFunction,
  evaluate,
  evaluateForAsync,
  // 其他
  LazyComponent,
  Overlay,
  PopOver,
  addSchemaFilter,
  OptionsControlProps,
  FormOptionsControl,
  FormControlProps,
  FormBaseControl,
  extendDefaultEnv,
  addRootWrapper,
  RendererConfig,
  styleMap,
  RENDERER_TRANSMISSION_OMIT_PROPS,
  ScopedComponentType,
  IItem,
  IColumn,
  IRow,
  IColumn2,
  IRow2,
  OnEventProps,
  FormSchemaBase,
  filterTarget,
  CustomStyle,
  enableDebug,
  disableDebug
};

export function render(
  schema: Schema,
  props: RootRenderProps = {},
  options: RenderOptions = {},
  pathPrefix: string = ''
): JSX.Element {
  return (
    <AMISRenderer
      {...props}
      schema={schema}
      pathPrefix={pathPrefix}
      options={options}
    />
  );
}

function AMISRenderer({
  schema,
  options,
  pathPrefix,
  ...props
}: RootRenderProps & {
  schema: Schema;
  options: RenderOptions;
  pathPrefix: string;
}) {
  let locale = props.locale || getDefaultLocale();
  // 兼容 locale 的不同写法
  locale =
    locale === 'en'
      ? 'en-US'
      : locale === 'zh' || locale === 'cn'
      ? 'zh-CN'
      : locale.replace('_', '-');

  const translate = React.useCallback(
    function () {
      const fn = props.translate || makeTranslator(locale);
      return fn.apply(null, arguments);
    },
    [locale, props.translate]
  );
  const store = React.useMemo(() => {
    let store = stores[options.session || 'global'];

    if (!store) {
      options = {
        ...defaultOptions,
        ...options,
        fetcher: options.fetcher
          ? wrapFetcher(options.fetcher, options.tracker)
          : defaultOptions.fetcher,
        confirm: promisify(
          options.confirm || defaultOptions.confirm || window.confirm
        ),
        locale,
        translate
      } as any;

      store = RendererStore.create({}, options);
      stores[options.session || 'global'] = store;
    } else {
      // 更新 env
      const env = getEnv(store);
      Object.assign(env, {
        ...options,
        fetcher: options.fetcher
          ? wrapFetcher(options.fetcher, options.tracker)
          : env.fetcher,
        confirm: options.confirm ? promisify(options.confirm) : env.confirm,
        locale,
        translate
      });
    }

    (window as any).amisStore = store; // 为了方便 debug.
    return store;
  }, Object.keys(options).concat(Object.values(options)).concat(locale));

  const env = getEnv(store);
  let theme = props.theme || options.theme || 'cxd';
  if (theme === 'default') {
    theme = 'cxd';
  }
  env.theme = getTheme(theme);

  React.useEffect(() => {
    env.enableAMISDebug ? enableDebug() : disableDebug();
    return () => env.enableAMISDebug || disableDebug();
  }, [env.enableAMISDebug]);

  if (props.locale !== undefined) {
    env.translate = translate;
    env.locale = locale;
  }

  // 默认将开启移动端原生 UI
  if (options.useMobileUI !== false) {
    props.mobileUI = env.isMobile();
  }

  // 根据环境覆盖 schema，这个要在最前面做，不然就无法覆盖 validations
  envOverwrite(schema, locale);
  schema = replaceText(schema, options.replaceText, env.replaceTextIgnoreKeys);

  return (
    <EnvContext.Provider value={env}>
      <ScopedRootRenderer
        {...props}
        schema={schema}
        pathPrefix={pathPrefix}
        rootStore={store}
        env={env}
        theme={theme}
        locale={locale}
        translate={translate}
      />
    </EnvContext.Provider>
  );
}
