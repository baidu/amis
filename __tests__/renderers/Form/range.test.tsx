import React = require('react');
import {render, fireEvent} from 'react-testing-library';
import '../../../src/themes/default';
import {render as amisRender} from '../../../src/index';
import {makeEnv} from '../../helper';

test('Renderer:number', async () => {
  const {container, getByRole} = render(
    amisRender(
      {
        type: 'form',
        api: '/api/xxx',
        controls: [
          {
            type: 'range',
            name: 'a',
            label: 'range',
            min: 0,
            max: 20,
            step: 2,
            value: 10,
            showInput: true
          }
        ],
        title: 'The form',
        actions: []
      },
      {},
      makeEnv({})
    )
  );

  fireEvent.mouseDown(getByRole('slider'));
  fireEvent.mouseMove(getByRole('slider'), {
    clientX: 400,
    clientY: 400
  });
  fireEvent.mouseUp(getByRole('slider'));

  const input = container.querySelector('input[name=a]');
  fireEvent.change(input!, {
    target: {
      value: '7'
    }
  });
  expect(container).toMatchSnapshot();
});

test('Renderer:range:multiple', async () => {
  const {container} = render(
    amisRender(
      {
        type: 'form',
        api: '/api/xxx',
        controls: [
          {
            type: 'range',
            name: 'a',
            label: 'range',
            min: 0,
            max: 20,
            step: 2,
            value: '10,15',
            multiple: true,
            joinValues: true,
            delimiter: ',',
            clearable: true,
            showInput: true
          }
        ],
        title: 'The form',
        actions: []
      },
      {},
      makeEnv({})
    )
  );

  const inputs = container.querySelectorAll('input[name=a]');
  fireEvent.change(inputs[0], {
    target: {
      value: '7'
    }
  });
  fireEvent.blur(inputs[0]);
  fireEvent.change(inputs[1], {
    target: {
      value: '13'
    }
  });
  fireEvent.blur(inputs[1]);
  const close = container.querySelector('a.a-InputRange-clear');
  fireEvent.click(close!);

  expect(container).toMatchSnapshot();
});
