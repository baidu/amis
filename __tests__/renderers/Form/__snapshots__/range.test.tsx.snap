// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer:number 1`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Form-item a-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  range
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-RangeControl a-RangeControl--withInput a-RangeControl--clearable a-Form-control"
          >
            <div
              aria-disabled="false"
              class="a-InputRange"
            >
              <span
                class="a-InputRange-label a-InputRange-label--min"
              >
                <span
                  class="a-InputRange-labelContainer"
                >
                  0
                </span>
              </span>
              <div
                class="a-InputRange-track a-InputRange-track--background"
              >
                <div
                  class="a-InputRange-track"
                  style="left: 0%; width: 40%;"
                />
                <span
                  class="a-InputRange-sliderContainer"
                  style="position: absolute; left: 40%;"
                >
                  <span
                    class="a-InputRange-label a-InputRange-label--value"
                    style="left: calc(50% - 0px);"
                  >
                    <span
                      class="a-InputRange-labelContainer"
                    >
                      8
                    </span>
                  </span>
                  <div
                    aria-valuemax="20"
                    aria-valuemin="0"
                    aria-valuenow="8"
                    class="a-InputRange-slider"
                    draggable="false"
                    role="slider"
                    tabindex="0"
                  />
                </span>
              </div>
              <span
                class="a-InputRange-label a-InputRange-label--max"
              >
                <span
                  class="a-InputRange-labelContainer"
                >
                  20
                </span>
              </span>
            </div>
            <span
              class="a-InputRange-label a-InputRange-label--mid"
              style="left: calc(50% - 60px);"
            >
              <span
                class="a-InputRange-labelContainer"
              >
                10
              </span>
            </span>
            <div
              class="a-InputRange-input"
            >
              <input
                class="is-active"
                name="a"
                type="text"
                value="8"
              />
            </div>
            <a
              class="a-InputRange-clear is-active"
            >
              <icon-mock
                classname="icon icon-close"
                icon="close"
              />
            </a>
          </div>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;

exports[`Renderer:range:multiple 1`] = `
<div>
  <div
    class="a-Panel a-Panel--default a-Panel--form"
    style="position: relative;"
  >
    <div
      class="a-Panel-heading"
    >
      <h3
        class="a-Panel-title"
      >
        <span
          class="a-TplField"
        >
          <span>
            The form
          </span>
        </span>
      </h3>
    </div>
    <div
      class="a-Panel-body"
    >
      <form
        class="a-Form a-Form--normal"
        novalidate=""
      >
        <div
          class="a-Form-item a-Form-item--normal"
          data-role="form-item"
        >
          <label
            class="a-Form-label"
          >
            <span>
              <span
                class="a-TplField"
              >
                <span>
                  range
                </span>
              </span>
            </span>
          </label>
          <div
            class="a-RangeControl a-RangeControl--withInput a-RangeControl--clearable is-multiple a-Form-control"
          >
            <div
              aria-disabled="false"
              class="a-InputRange"
            >
              <span
                class="a-InputRange-label a-InputRange-label--min"
              >
                <span
                  class="a-InputRange-labelContainer"
                >
                  0
                </span>
              </span>
              <div
                class="a-InputRange-track a-InputRange-track--background"
              >
                <div
                  class="a-InputRange-track is-active"
                  style="left: 0%; width: 100%;"
                />
                <span
                  class="a-InputRange-sliderContainer"
                  style="position: absolute; left: 0%;"
                >
                  <span
                    class="a-InputRange-label a-InputRange-label--value"
                  >
                    <span
                      class="a-InputRange-labelContainer"
                    >
                      0
                    </span>
                  </span>
                  <div
                    aria-valuemax="20"
                    aria-valuemin="0"
                    aria-valuenow="0"
                    class="a-InputRange-slider"
                    draggable="false"
                    role="slider"
                    tabindex="0"
                  />
                </span>
                <span
                  class="a-InputRange-sliderContainer"
                  style="position: absolute; left: 100%;"
                >
                  <span
                    class="a-InputRange-label a-InputRange-label--value"
                  >
                    <span
                      class="a-InputRange-labelContainer"
                    >
                      20
                    </span>
                  </span>
                  <div
                    aria-valuemax="20"
                    aria-valuemin="0"
                    aria-valuenow="20"
                    class="a-InputRange-slider"
                    draggable="false"
                    role="slider"
                    tabindex="0"
                  />
                </span>
              </div>
              <span
                class="a-InputRange-label a-InputRange-label--max"
              >
                <span
                  class="a-InputRange-labelContainer"
                >
                  20
                </span>
              </span>
            </div>
            <span
              class="a-InputRange-label a-InputRange-label--mid"
              style="left: calc(50% - 60px);"
            >
              <span
                class="a-InputRange-labelContainer"
              >
                10
              </span>
            </span>
            <div
              class="a-InputRange-input is-multiple"
            >
              <input
                class=""
                name="a"
                type="text"
                value="0"
              />
              <span
                class="a-InputRange-input-separator"
              >
                 - 
              </span>
              <input
                class=""
                name="a"
                type="text"
                value="20"
              />
            </div>
            <a
              class="a-InputRange-clear is-active"
            >
              <icon-mock
                classname="icon icon-close"
                icon="close"
              />
            </a>
          </div>
        </div>
        <input
          style="display: none;"
          type="submit"
        />
      </form>
    </div>
    <div
      class="resize-sensor"
      style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: scroll; z-index: -1; visibility: hidden;"
    >
      
  
      <div
        class="resize-sensor-expand"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0px; top: 0px; width: 10px; height: 10px;"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-shrink"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;"
      >
        
    
        <div
          style="position: absolute; left: 0; top: 0; width: 200%; height: 200%"
        />
        
  
      </div>
      
  
      <div
        class="resize-sensor-appear"
        style="position: absolute; left: 0; top: 0; right: 0; bottom: 0; overflow: scroll; z-index: -1; visibility: hidden;animation-name: apearSensor; animation-duration: 0.2s;"
      />
    </div>
  </div>
</div>
`;
